flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java

```diff
@@ -71,22 +94,67 @@ public boolean contains(UK key) throws Exception {
 
     @Override
     public Iterable<Map.Entry<UK, UV>> entries() throws Exception {
-        return delegatedState.entries();
+        Iterator<Map.Entry<UK, UV>> iterator = delegatedState.entries().iterator();
+        final N currentNamespace = getCurrentNamespace();
+        return () ->
+                loggingIterator(
+                        new Iterator<Map.Entry<UK, UV>>() {
+                            @Override
+                            public Map.Entry<UK, UV> next() {
+                                return loggingMapEntry(
+                                        iterator.next(),
+                                        changeLogger,
+                                        changeWriter,
+                                        currentNamespace);
+                            }
+
+                            @Override
+                            public boolean hasNext() {
+                                return iterator.hasNext();
+                            }
+
+                            @Override
+                            public void remove() {
+                                iterator.remove();
+                            }
+                        },
+                        changeLogger,
+                        changeWriter,
+                        currentNamespace);
     }
 
     @Override
     public Iterable<UK> keys() throws Exception {
-        return delegatedState.keys();
+        return loggingIterable(
+                delegatedState.keys(), changeLogger, this::serializeKey, getCurrentNamespace());
```

Comment:
Is it possible to remove a key without removing its value for a map?
Ask because the loggingIterable is mainly for remove change

Reply:
No, it's impossible. I'm not sure I understand your concern though.

What can happen is:
1. user calls `map.put(1, 1)`
2. `delegatedMap` creates a new `HashMap` for the current stream key, stores it, and calls `put(1, 1)` on it
3. user calls `changeLogMap.keys()`
4. `delegatedMap` returns `hashMap.keySet`, wrapped for logging `ChangelogMapState`
5. user calls `keys.iterator.next().remove(x)` - this is supported by jdk `HashMap` and should be logged. Both key and value are removed.

Reply:
Let me rephrase my question:

1. Some backend supports iterator change/remove; but some does not, but this is not the focus, I am asking in the case where iterator change/remove are supported (rocksdb for example)

2. In the case where iterator is supported
Let's say I get an iterator list of keys, and I get an iterator list of values, I remove some keys in the key iterator, and some values in the value iterator

Is it guaranteed that such remove is always paired? (Key, Value)